drivers/flash/flash_mspi_infineon_s25h.c

```diff
@@ -0,0 +1,836 @@
+	uint8_t val = 0;
+
+	ret = flash_mspi_infineon_s25h_rw_any_register(dev, INF_MSPI_S25H_ADDRESS_VOLATILE_STATUS_1,
+						       &val, 0, MSPI_RX);
+	if (ret < 0) {
+		return ret;
+	}
+	*is_enabled = (val & INF_MSPI_S25H_STATUS_1_WRPGEN_BIT);
+	return 0;
+}
+
+static int flash_mspi_infineon_s25h_wait_for_idle(const struct device *dev, uint32_t timeout_ms)
+{
+	int ret = 0;
+	uint8_t status_1 = 0;
+	uint32_t retries = timeout_ms / INF_MSPI_S25H_TIMEOUT_IDLE_RETRY_INTERVAL_MS;
+
+	ret = flash_mspi_infineon_s25h_rw_any_register(dev, INF_MSPI_S25H_ADDRESS_VOLATILE_STATUS_1,
+						       &status_1, 0, MSPI_RX);
+	if (ret < 0) {
+		return ret;
+	}
+
+	while ((status_1 & INF_MSPI_S25H_STATUS_1_RDYBSY_BIT) != 0 && retries > 0) {
+		k_sleep(INF_MSPI_S25H_TIMEOUT_IDLE_RETRY_INTERVAL);
+		ret = flash_mspi_infineon_s25h_rw_any_register(
+			dev, INF_MSPI_S25H_ADDRESS_VOLATILE_STATUS_1, &status_1, 0, MSPI_RX);
+		if (ret < 0) {
+			return ret;
+		}
+		--retries;
+	}
+
+	if (retries == 0) {
+		LOG_ERR("Waiting for flash to enter idle. Status 1 register: 0x%X", status_1);
+		return -EIO;
+	}
+
+	return 0;
+}
+
+static int flash_mspi_infineon_s25h_read_jedec_id(const struct device *dev, uint8_t *buf)
+{
+	int ret = 0;
+	const struct flash_mspi_infineon_s25h_cfg *config = dev->config;
+	struct flash_mspi_infineon_s25h_data *data = dev->data;
+
+	const struct mspi_xfer_packet packet = {
+		.dir = MSPI_RX,
+		.cmd = data->read_jedec_cmd,
+		.num_bytes = 3,
+		.data_buf = buf,
+		.address = 0,
+	};
+
+	struct mspi_xfer xfer = {
+		INF_MSPI_S25H_DEFAULT_XFER_DATA,
+		.addr_length = 0,
+		.rx_dummy = 0,
+		.packets = &packet,
+		.num_packet = 1,
+		.timeout = INF_MSPI_S25H_DEFAULT_MSPI_TIMEOUT,
+	};
+
+	ret = mspi_transceive(config->bus, &config->dev_id, &xfer);
+	if (ret < 0) {
+		LOG_ERR("Error reading JEDEC id");
+		return ret;
+	}
+
+	return 0;
+}
+
+static int flash_mspi_infineon_s25h_read(const struct device *dev, off_t addr, void *data,
+					 size_t size)
+{
+	int ret = 0;
+	bool requires_cleanup = false;
+	const struct flash_mspi_infineon_s25h_cfg *config = dev->config;
+	struct flash_mspi_infineon_s25h_data *dev_data = dev->data;
+
+	/* The S25H allows for continuous read operations which happen by sending
+	 * 0xAX after an address. The driver doesn't implement this which is why we
+	 * don't want this and instead wait for 2 cycles. However since the flash
+	 * pins could be in high impedance state from the MSPI controller after the
+	 * address was sent an address ending with 0xA could put the flash into a
+	 * continuous read mode. To prevent this the driver will read the jedec id,
+	 * if the address wasn't aligned to prevent accidentally fulfilling the
+	 * requirement.
+	 */
+	if (dev_data->mspi_dev_cfg.io_mode == MSPI_IO_MODE_QUAD && (addr % 16 != 0)) {
+		requires_cleanup = true;
+	}
+
+	ret = flash_mspi_infineon_s25h_prepare_mspi_bus(dev);
+	if (ret < 0) {
+		LOG_ERR("Error setting up the MSPI bus for the flash device");
+		return ret;
+	}
```

Comment:
Is there a need to configure the controller for every read? I would assume the controller configuration stays the same. If needed, the application can explicitly configure the controller. Same with the `write` API.